### PR TITLE
Troposphere Delay Model

### PR DESCRIPTION
**Pull request by zimmyneutron**
This adds a model to account for the tropospheric delay of TOAs for ground-based observatory.  It will use the location of each observatory to calculate the angular altitude of the target pulsar and corresponding delay.  

The models used are:

- Zenith hydrostatic delay based on [Davis et al. (1985)](https://agupubs.onlinelibrary.wiley.com/doi/abs/10.1029/RS020i006p01593)
- Mapping function to different locations in the sky from [Niell (1996)](https://www.haystack.mit.edu/geo/pubs/NMF_JGR.pdf)
- Eventual inclusion of wet zenith delay and corresponding Niell wet mapping function

### CODE REVIEW

**Comment by paulray** [comment]
Looks like you are fighting with getting the path for data files.  I think the preferred way to do that is like this:
```
from pinttestdata import testdir

datadir = os.path.join(testdir, "datafile")
parfile = os.path.join(datadir, "J1744-1134.basic.par")
```
The tests should not make an assumption about the current working directory when you run them. Sometime I run them as `pytest tests/test_xxx.py` and other times as `pytest test_xxx.py` and they should work either way.

**Comment by zimmyneutron** [comment]
Thanks I'll put in that change

File: src/pint/models/troposphere_delay.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+    def _get_target_altitude(self, obs, grp, radec):
+        """convert the sky coordinates of the target to the angular altitude at each TOA
+        """
+        transformAltaz = AltAz(location=obs, obstime=grp["mjd"])
+        alt = radec.transform_to(transformAltaz).alt  # * u.deg
+        return alt
+
+    def _get_target_skycoord(self):
+        """return the sky coordiantes for the target, either from equatorial or ecliptic coordinates
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:129]: typo in "coordiantes" spelling

File: src/pint/models/troposphere_delay.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        """
+        gph = self.EARTH_R * H / (self.EARTH_R + H)  # geopotential height
+        if gph > 11 * u.km:
+            warn("Pressure approximation invalid for elevations above 11 km")
+        T = 288.15 - 0.0065 * H.to(u.m).value  # temperature lapse
+        P = 101.325 * (288.15 / T) ** -5.25575 * u.kPa
+        return P
+
+    def zenith_delay(self, lat, H):
+        """Calculate the hydrostatic zenith delay
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:246]: Should have a reference where this equation is coming from

File: src/pint/models/troposphere_delay.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        """
+        p = self.pressure_from_altitude(H)
+        return (p / (43.921 * u.kPa)) / (
+            const.c.value * (1 - 0.00266 * np.cos(2 * lat) - 0.00028 * H.value)
+        )
+
+    def wet_zenith_delay(self):
+        """ calculate the wet delay at zentih
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:254]: "zentih"

File: src/pint/models/troposphere_delay.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        """
+        absLat = np.abs(lat)
+        for lInd in range(len(self.LAT)):
+            if absLat >= self.LAT[lInd - 1]:
+                return lInd
+        # else this is an invalid latitude... huh?
+        raise ValueError("Invaid latitude: %s must be between -90 and 90 degrees" % lat)
+
+    def mapping_function(self, alt, lat, H, mjd):
+        """this implements the niell mapping function for hydrostatic
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:277]: capitalize niell

File: src/pint/models/troposphere_delay.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    )
+
+    LAT = np.array([0, 15, 30, 45, 60, 75, 90]) * u.deg  # in degrees
+
+    DOY_OFFSET = -28  # add this into the MJD value to get the right phase
+
+    EARTH_R = 6356766 * u.m  # earth radius at 45 degree latitude
+
+    @staticmethod
+    def _herring_map(alt, a, b, c):
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:78]: What exactly is this?  Might be good to describe it in a sentence or two.

File: src/pint/models/troposphere_delay.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+from pint.observatory.topo_obs import TopoObs
+from pint.toa_select import TOASelect
+
+
+class TroposphereDelay(DelayComponent):
+    """
+
+    Model for accounting for the troposphere delay for topocentric TOAs.
+
+    Based on Davis zenith hydrostatic delay (Davis et al., 1985, Appendix A)
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:24]: Can you explain which parts of the model come from which of the 2 references?

File: src/pint/models/troposphere_delay.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+
+        # now add in the mapping correction based on height
+        fcorrection = self._herring_map(alt, self.A_HT, self.B_HT, self.C_HT)
+
+        return baseMap + (1 / np.sin(alt) - fcorrection) * H.to(u.km).value
+
+    def wet_map(self, alt, lat):
+        """this is very similar to the normal mapping function except it uses different coefficients
+        and i believe that there isn't an explicit height correction to the mapping function,
+        but i'll double check the literature.
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:322]: Did you find out about the explicit height correction in the literature?

File: src/pint/models/troposphere_delay.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+        latNeighbors = self.LAT[latIndex : latIndex + 2]
+
+        a = self._interp(np.abs(lat), latNeighbors, aNeighbors)
+        b = self._interp(np.abs(lat), latNeighbors, bNeighbors)
+        c = self._interp(np.abs(lat), latNeighbors, cNeighbors)
+
+        return self._herring_map(alt, a, b, c)
+
+    @staticmethod
+    def _interp(x, xn, yn):
```
Comment (scottransom) [inline src/pint/models/troposphere_delay.py:340]: Can you not using one of the numpy interpolation methods rather than implementing your own?

Comment (zimmyneutron) [reply src/pint/models/troposphere_delay.py:340]: Ideally I would, but I'm having trouble getting the numpy/scipy interpolation to work with vectorized quantities.  For a TOA observed at 35 degree latitude, I calculate the mapping function at 30 degrees and 45 degrees using the table of values.  This depends on the time of year, so it is a vectorized calculation for each TOA at that observatory.  Then I interpolate between the two results to the correct latitude.  And that's where I get the issues, with numpy not recognizing the vector structure properly and crashing.  I can continue trying to get it to work if it's necessary to use the built-in library for this calculation.

**Comment by scottransom** [comment]
The ModelBuilder is not finding the troposphere_delay model component.  You need to import the class in __init__.py in the models directory.  I think ModelBuilder can figure it out after that.

````
diff --git a/src/pint/models/__init__.py b/src/pint/models/__init__.py
index 39abd3d..3d06457 100644
--- a/src/pint/models/__init__.py
+++ b/src/pint/models/__init__.py
@@ -16,6 +16,7 @@ from pint.models.noise_model import EcorrNoise, PLRedNoise, ScaleToaError
 from pint.models.solar_system_shapiro import SolarSystemShapiro
 from pint.models.solar_wind_dispersion import SolarWindDispersion
 from pint.models.spindown import Spindown
+from pint.models.troposphere_delay import TroposphereDelay
 
 # Import the main timing model classes
 from pint.models.timing_model import TimingModel, DEFAULT_ORDER
```

**Comment by scottransom** [comment]
One nice little change that would be good would be to make the model pretty printer handle CORRECT_TROPOSPHERE better.  Here is what it looks like now:
```

PAR                   Prefit                  Postfit            Units
============== ==================== ============================ =====
PSR                      1748-2021E                              None
EPHEM                         DE421                              None
UNITS                           TDB                              None
POSEPOCH                      53750                              d 
PX                                0                              mas 
RAJ                    17h48m52.75s    17h48m52.8003s +/- 0.00014 hourangle_second 
DECJ                     -20d21m29s   -20d21m29.3833s +/- 0.033 arcsec 
PMRA                              0                              mas / yr 
PMDEC                             0                              mas / yr 
F0                          61.4855          61.485476554373(18) Hz 
F1                       -1.181e-15            -1.1813(14)×10⁻¹⁵ Hz / s 
PEPOCH                        53750                              d 
CORRECT_TROPOSPHERE                    0                              None 
PLANET_SHAPIRO                    0                              None 
NE_SW                             0                              1 / cm3 
SWM                               0                               
DM                            223.9                  224.114(35) pc / cm3 
DM1                               0                              pc / (cm3 yr) 
TZRMJD                      53801.4                              d 
TZRSITE                           1                              None
TZRFRQ                      1949.61                              MHz 
```

**Comment by paulray** [comment]
Since you are updating the AUTHORS.rst list, would you mind moving Chloe up from other contributors to the paper authors section?

**Comment by scottransom** [comment]
@paulray As soon as Travis checks finish I'm going to merge this.  Feel free to just push up a direct bug fix for the AUTHORS.rst file (since that can't break PINT by definition!)

**Comment by zimmyneutron** [comment]
I actually just updated it and pushed the change

**Comment by zimmyneutron** [comment]
@scottransom Has the TroposphereDelay passed all your checks to merge?

**Pull request merged**